### ISSUE TITLE
No response from invoke container for Lambda inside docker-compose

### ISSUE THREAD

**Issue by andrewx82** [aws/aws-sam-cli#2837]
### Description:

Related to https://github.com/aws/aws-sam-cli/issues/2492

When running a docker-compose with an image that is using sam to run a local api, the api can never get a response from the lambda container. Everything works correctly when run locally not in a container.

I have read through all the answers in 2492 and tried the various permutations suggested, nothing has worked. There appeared to be no subsequent documentation follow up for getting this setup to work with docker.

### Steps to reproduce:

DockerFile

```
FROM python:alpine
RUN apk add --no-cache --virtual build-deps build-base=0.5-r2 gcc=10.2.1_pre1-r3 bash=5.1.0-r0 && \
    pip install aws-sam-cli==1.23.0 && \
    apk del build-deps
COPY ./sam_entrypoint.sh /bin/sam_entrypoint.sh
RUN chmod +x /bin/sam_entrypoint.sh

WORKDIR /app
COPY ./node_modules /app/node_modules
COPY ./package.json /app/package.json
COPY ./template.yml /app/template.yml
COPY ./dist /app/dist
EXPOSE 3000 
ENTRYPOINT ["sh","/bin/sam_entrypoint.sh"]
```

sam_entrypoint.sh
```
#!/bin/bash
set -o errexit
/usr/local/bin/sam local start-api --debug \
  --template ./template.yml \
  --host 0.0.0.0 \
  --container-host host.docker.internal \
  --container-host-interface  127.0.0.1 \
  --docker-network host
````

Template.yml
```
AWSTemplateFormatVersion: "2010-09-09"
Transform: "AWS::Serverless-2016-10-31"
Description: "Automated lambda setup for api gateway"
Resources:
  res0:
    Type: "AWS::Serverless::Function"
    Properties:
      CodeUri: "./"
      Handler: "dist/index.handler"
      Runtime: "nodejs12.x"
      Timeout: 10
      Events:
        CatchAll:
          Type: "Api"
          Properties:
            Path: "/"
            Method: ANY
  res1:
    Type: "AWS::Serverless::Function"
    Properties:
      CodeUri: "./"
      Handler: "dist/index.handler"
      Runtime: "nodejs12.x"
      Timeout: 10
      Events:
        CatchAll:
          Type: "Api"
          Properties:
            Path: "/{proxy+}"
            Method: ANY
```

Docker-compose 
```
version: '3.6'
services:
  sam_app:
    build: 
      context: .
      dockerfile: Dockerfile.test
    ports:
      - "3000:3000"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock  
    extra_hosts:
      - "host.docker.internal:host-gateway"
```

### Observed result:

```
sam_app_1  | 
sam_app_1  |    SAM CLI now collects telemetry to better understand customer needs.
sam_app_1  |
sam_app_1  |    You can OPT OUT and disable telemetry collection by setting the
sam_app_1  |    environment variable SAM_CLI_TELEMETRY=0 in your shell.
sam_app_1  |    Thanks for your help!
sam_app_1  |
sam_app_1  |    Learn More: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-telemetry.html
sam_app_1  |
sam_app_1  | 2021-04-28 17:40:50,630 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
sam_app_1  | 2021-04-28 17:40:50,630 | Using config file: samconfig.toml, config environment: default
sam_app_1  | 2021-04-28 17:40:50,631 | Expand command line arguments to:
sam_app_1  | 2021-04-28 17:40:50,631 | --template_file=/app/template.yml --host=0.0.0.0 --port=3000 --static_dir=public --layer_cache_basedir=/root/.aws-sam/layers-pkg --container_host=localhost --container_host_interface=127.0.0.1      
sam_app_1  | 2021-04-28 17:40:50,704 | local start-api command is called
sam_app_1  | 2021-04-28 17:40:50,709 | No Parameters detected in the template
sam_app_1  | 2021-04-28 17:40:50,732 | 3 stacks found in the template
sam_app_1  | 2021-04-28 17:40:50,732 | No Parameters detected in the template
sam_app_1  | 2021-04-28 17:40:50,753 | 3 resources found in the stack
sam_app_1  | 2021-04-28 17:40:50,753 | No Parameters detected in the template
sam_app_1  | 2021-04-28 17:40:50,773 | Found Serverless function with name='res0' and CodeUri='./'
sam_app_1  | 2021-04-28 17:40:50,773 | --base-dir is not presented, adjusting uri ./ relative to /app/template.yml
sam_app_1  | 2021-04-28 17:40:50,773 | Found Serverless function with name='res1' and CodeUri='./'
sam_app_1  | 2021-04-28 17:40:50,773 | --base-dir is not presented, adjusting uri ./ relative to /app/template.yml
sam_app_1  | 2021-04-28 17:40:50,777 | No Parameters detected in the template
sam_app_1  | 2021-04-28 17:40:50,798 | No Parameters detected in the template
sam_app_1  | 2021-04-28 17:40:50,819 | Found '1' API Events in Serverless function with name 'res0'
sam_app_1  | 2021-04-28 17:40:50,819 | Found '1' API Events in Serverless function with name 'res1'
sam_app_1  | 2021-04-28 17:40:50,819 | Detected Inline Swagger definition
sam_app_1  | 2021-04-28 17:40:50,819 | Lambda function integration not found in Swagger document at path='/' method='x-amazon-apigateway-any-method'
sam_app_1  | 2021-04-28 17:40:50,819 | Lambda function integration not found in Swagger document at path='/{proxy+}' method='x-amazon-apigateway-any-method'
sam_app_1  | 2021-04-28 17:40:50,819 | Found '0' APIs in resource 'ServerlessRestApi'
sam_app_1  | 2021-04-28 17:40:50,819 | Removed duplicates from '0' Explicit APIs and '2' Implicit APIs to produce '2' APIs
sam_app_1  | 2021-04-28 17:40:50,819 | 2 APIs found in the template
sam_app_1  | 2021-04-28 17:40:50,825 | Mounting res0 at http://0.0.0.0:3000/ [DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT]
sam_app_1  | 2021-04-28 17:40:50,825 | Mounting res1 at http://0.0.0.0:3000/{proxy+} [DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT]
sam_app_1  | 2021-04-28 17:40:50,826 | You can now browse to the above endpoints to invoke your functions. You do not need to restart/reload SAM CLI while working on your functions, changes will be reflected instantly/automatically. You 
only need to restart SAM CLI if you update your AWS SAM template
sam_app_1  | 2021-04-28 17:40:50,826 | Localhost server is starting up. Multi-threading = True
sam_app_1  | 2021-04-28 17:40:50  * Running on http://0.0.0.0:3000/ (Press CTRL+C to quit)
sam_app_1  | 2021-04-28 17:41:20,791 | Constructed String representation of Event to invoke Lambda. Event: {"body": null, "headers": {"Accept": "*/*", "Host": "127.0.0.1:3000", "User-Agent": "insomnia/2021.2.2", "X-Forwarded-Port": "3000", "X-Forwarded-Proto": "http"}, "httpMethod": "GET", "isBase64Encoded": false, "multiValueHeaders": {"Accept": ["*/*"], "Host": ["127.0.0.1:3000"], "User-Agent": ["insomnia/2021.2.2"], "X-Forwarded-Port": ["3000"], "X-Forwarded-Proto": 
["http"]}, "multiValueQueryStringParameters": null, "path": "/v1/openapi.json", "pathParameters": {"proxy": "v1/openapi.json"}, "queryStringParameters": null, "requestContext": {"accountId": "123456789012", "apiId": "1234567890", "domainName": "127.0.0.1:3000", "extendedRequestId": null, "httpMethod": "GET", "identity": {"accountId": null, "apiKey": null, "caller": null, "cognitoAuthenticationProvider": null, "cognitoAuthenticationType": null, "cognitoIdentityPoolId": null, "sourceIp": "172.24.0.1", "user": null, "userAgent": "Custom User Agent String", "userArn": null}, "path": "/{proxy+}", "protocol": "HTTP/1.1", "requestId": "d9c5fa20-2328-43f7-b4d2-5a35028dfab1", "requestTime": "28/Apr/2021:17:40:50 +0000", "requestTimeEpoch": 1619631650, "resourceId": "123456", "resourcePath": "/{proxy+}", "stage": "Prod"}, "resource": "/{proxy+}", "stageVariables": null, "version": "1.0"}
sam_app_1  | 2021-04-28 17:41:20,791 | Found one Lambda function with name 'res1'
sam_app_1  | 2021-04-28 17:41:20,791 | Invoking dist/index.handler (nodejs12.x)
sam_app_1  | 2021-04-28 17:41:20,791 | Environment variables overrides data is standard format
sam_app_1  | 2021-04-28 17:41:20,791 | Loading AWS credentials from session with profile 'None'
sam_app_1  | 2021-04-28 17:41:20,803 | Resolving code path. Cwd=/app, CodeUri=/app
sam_app_1  | 2021-04-28 17:41:20,803 | Resolved absolute path to code is /app
sam_app_1  | 2021-04-28 17:41:20,803 | Code /app is not a zip/jar file
sam_app_1  | 2021-04-28 17:41:20,817 | Skip pulling image and use local one: amazon/aws-sam-cli-emulation-image-nodejs12.x:rapid-1.23.0.
sam_app_1  |
sam_app_1  | 2021-04-28 17:41:20,817 | Mounting /app as /var/task:ro,delegated inside runtime container
sam_app_1  | 2021-04-28 17:41:21,124 | Starting a timer for 10 seconds for function 'res1'
sam_app_1  | 2021-04-28 17:41:22,075 | Cleaning all decompressed code dirs
sam_app_1  | 2021-04-28 17:41:22,075 | No response from invoke container for res1
sam_app_1  | 2021-04-28 17:41:22,075 | Invalid lambda response received: Lambda response must be valid json
sam_app_1  | 2021-04-28 17:41:22 172.24.0.1 - - [28/Apr/2021 17:41:22] "GET /v1/openapi.json HTTP/1.1" 502 -
```

### Expected result:
That the lambda can be executed correctly and a response received.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS:  Windows 10
2. `sam --version`: 1.23.0
3. AWS region: us-east-1

**Comment by hoffa** [comment]
Is your Docker daemon running locally? Could you try with `--container-host-interface 0.0.0.0` (which will bind the container ports to all network interfaces) to ensure it's not an issue with port binding?

**Comment by andrewx82** [comment]
Hi @hoffa I've tested with --container-host-interface 0.0.0.0, i still get no response from the container

**Comment by xazhao** [comment]
Hi @andrewx82 could you try the command without `--docker-network` option? The default bridge network should work for your case.

**Comment by andrewx82** [comment]
Hi @xazhao The call is now accessible and I see it being processed in the logs , however I see no response from the containers
```
sam_app_1  | 2021-05-04 11:43:05,155 | Found one Lambda function with name 'res1'
sam_app_1  | 2021-05-04 11:43:05,155 | Invoking dist/index.handler (nodejs12.x)
sam_app_1  | 2021-05-04 11:43:05,155 | Environment variables overrides data is standard format
sam_app_1  | 2021-05-04 11:43:05,155 | Loading AWS credentials from session with profile 'None'
sam_app_1  | 2021-05-04 11:43:05,168 | Resolving code path. Cwd=/app, CodeUri=/app
sam_app_1  | 2021-05-04 11:43:05,168 | Resolved absolute path to code is /app
sam_app_1  | 2021-05-04 11:43:05,169 | Code /app is not a zip/jar file
sam_app_1  | 2021-05-04 11:43:05,173 | Image was not found.
sam_app_1  | Building image...........................................................................................................................................................................................................................................................................................................................................................
sam_app_1  | 2021-05-04 11:44:41,985 | Skip pulling image and use local one: amazon/aws-sam-cli-emulation-image-nodejs12.x:rapid-1.23.0.
sam_app_1  |
sam_app_1  | 2021-05-04 11:44:41,985 | Mounting /app as /var/task:ro,delegated inside runtime container
sam_app_1  | 2021-05-04 11:44:42,334 | Starting a timer for 10 seconds for function 'res1'
sam_app_1  | 2021-05-04 11:44:43,246 | Cleaning all decompressed code dirs
sam_app_1  | 2021-05-04 11:44:43,246 | No response from invoke container for res1
sam_app_1  | 2021-05-04 11:44:43,247 | Invalid lambda response received: Lambda response must be valid json
sam_app_1  | 2021-05-04 11:44:43 172.20.0.1 - - [04/May/2021 11:44:43] "GET /v1/openapi.json HTTP/1.1" 502 -
```

**Comment by xazhao** [comment]
@andrewx82 `extra-host` in your docker compose file is for Linux. Since your OS is Windows, I guess that is actually causing the issue. Can you try it without `extra-host`?

**Comment by andrewx82** [comment]
Hi @xazhao 

I have changed the sam_entrypoint.sh to the following 
```#!/bin/bash
set -o errexit
BASEDIR=$(pwd)
/usr/local/bin/sam local start-api --debug \
  --template ./template.yml \
  --host 0.0.0.0 \
  # --docker-volume-basedir "${BASEDIR}" \
  --container-host host.docker.internal \
  # --container-host-interface  0.0.0.0 \
  # --docker-network host
```

along with the docker-compose to 

```
version: '3.6'
services:
  sam_app:
    build: 
      context: .
      dockerfile: Dockerfile.test
    ports:
      - "3000:3000"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock  
```

Still appears to be no response


```
sam_app_1  | 2021-05-05 07:47:59,514 | Found one Lambda function with name 'res1'
sam_app_1  | 2021-05-05 07:47:59,515 | Invoking dist/index.handler (nodejs12.x)
sam_app_1  | 2021-05-05 07:47:59,515 | Environment variables overrides data is standard format
sam_app_1  | 2021-05-05 07:47:59,515 | Loading AWS credentials from session with profile 'None'
sam_app_1  | 2021-05-05 07:47:59,531 | Resolving code path. Cwd=/app, CodeUri=/app
sam_app_1  | 2021-05-05 07:47:59,531 | Resolved absolute path to code is /app
sam_app_1  | 2021-05-05 07:47:59,531 | Code /app is not a zip/jar file
sam_app_1  | 2021-05-05 07:47:59,547 | Skip pulling image and use local one: amazon/aws-sam-cli-emulation-image-nodejs12.x:rapid-1.23.0.
sam_app_1  |
sam_app_1  | 2021-05-05 07:47:59,547 | Mounting /app as /var/task:ro,delegated inside runtime container
sam_app_1  | 2021-05-05 07:47:59,823 | Starting a timer for 10 seconds for function 'res1'
sam_app_1  | 2021-05-05 07:48:00,764 | Cleaning all decompressed code dirs
sam_app_1  | 2021-05-05 07:48:00,765 | No response from invoke container for res1
sam_app_1  | 2021-05-05 07:48:00,765 | Invalid lambda response received: Lambda response must be valid json
sam_app_1  | 2021-05-05 07:48:00 172.20.0.1 - - [05/May/2021 07:48:00] "GET /v1/openapi.json HTTP/1.1" 502 -
```

**Comment by xazhao** [comment]
@andrewx82 I was testing using configurations from this comment: https://github.com/aws/aws-sam-cli/issues/2436#issuecomment-788317898. After adding `--container-host host.docker.internal` it works. It looks pretty similar setup with your setup though.

**Comment by andrewx82** [comment]
@xazhao I can confirm that that does not work . See my bash entry script 

```
#!/bin/bash
set -o errexit
BASEDIR=$(pwd)
/usr/local/bin/sam local start-api --debug \
  --template ./template.yml \
  -v $BASEDIR \
  --host 0.0.0.0 \
  --container-host host.docker.internal 
```

I also tried with and without the -v option

Can we just get an official example instead of  saying that it is supposed to work ?

**Comment by xazhao** [comment]
@andrewx82 This is what I'm currently using.
`Dockerfile`
```
FROM amazon/aws-cli

RUN yum -y update && \
    yum install -y python3 && \
    python3 -m pip install aws-sam-cli

ENTRYPOINT [ "" ]
```
`docker-compose.yml`
```
services:
  testcase:
    build: .
    working_dir: $PWD
    command: sam local start-api --host 0.0.0.0 --container-host host.docker.internal
    ports:
      - 3000:3000
    volumes:
      - '$PWD:$PWD'
      - '/var/run/docker.sock:/var/run/docker.sock'`
``` 
Then open another terminal and use the following command:
`curl localhost:3000/hello`

Please note: My development environment is macOS and I think mounting volume is different in Windows (Seems you already did properly). For your `No response ...` error, it usually happens when lambda function is not hosted on a container correctly. 
One thing you might try is see if you can resolve `host.docker.internal` to an IP address. For me, `host.docker.internal` resolves to `192.168.65.2`. If I use this ip instead of `host.docker.internal`, it will work too.

**Comment by andrewx82** [comment]
@xazhao  I've tried to align my files to be similar to yours
[aws-sam.zip](https://github.com/aws/aws-sam-cli/files/6507349/aws-sam.zip)

It appears that something is happening here as I'm getting information about interacting with the lambda, however it doesn't seem like its reading the values correctly. For clarity I don't want to use volumes for this as I want to bake a docker image with the lambda

**Comment by sbhvt** [comment]
I'm having a very similar issue. I think that the root cause is related to the code path it is attempting to mount. 

In the output that @andrewx82 posted earlier, you see this line: `Mounting /app as /var/task:ro,delegated inside runtime container`.  '/app' is the WORKDIR and is the path within the container but it is not the path on the host, and the docker engine needs the path on host for mounting. 

In order to properly evaluate paths for mounts, the docker engine needs the path on the host of the docker engine. In @xazhao 's examples, this is accomplished because of the `working_dir: $PWD` line in the docker-compose file. Using the sample code that @xaxhao shared, I am able to reproduce the same error that @andrewx82 shared by commenting out the `working_dir: $PWD` line from docker-compose.  

(Note that if using the exact code shared by @xazhao , the WORKDIR is "aws" instead of "app", but the issue is the same regardless.) Here's what you see logged when `working_dir: $PWD` is commented out:

```
sam_app_1  | 2021-05-20 20:26:52,932 | Resolving code path. Cwd=/aws, CodeUri=/aws/hello_world
sam_app_1  | 2021-05-20 20:26:52,932 | Resolved absolute path to code is /aws/hello_world
sam_app_1  | 2021-05-20 20:26:52,933 | Code /aws/hello_world is not a zip/jar file
...
```

In the use case of running from within VSCode devcontainers, which is one of the use cases mentioned in #2492, the option with `working_dir: $PWD` is not a feasible solution. 

I also tried this (working_dir commented out but passing $PWD to the `--docker-volume-basedir` parameter) :

```
version: '3.6'
services:
  sam_app:
    #working_dir: $PWD
    build:
      context: .
      dockerfile: Dockerfile
    ports:
      - "3000:3000"
    command: sam local start-api --host 0.0.0.0 --container-host host.docker.internal --debug --docker-volume-basedir $PWD
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - $PWD:$PWD
```

In that case, here is the the output:

```
sam_app_1  | 2021-05-20 20:30:33,580 | --base-dir is not presented, adjusting uri hello_world/ relative to /aws/template.yml
...
sam_app_1  | 2021-05-20 20:26:52,932 | Resolving code path. Cwd=<MY_LOCAL_PATH>, CodeUri=/aws/hello_world
sam_app_1  | 2021-05-20 20:26:52,932 | Resolved absolute path to code is /aws/hello_world
sam_app_1  | 2021-05-20 20:26:52,933 | Code /aws/hello_world is not a zip/jar file
...
```

I looked at the codebase related to that area of code: `resolve_code_path` in *samcli/lib/utils/codeuri.py*). The calling code already resolved the codeuri parameter to an absolute path before calling that method. Judging by the output, it looks like this happens in when it logs `--base-dir is not presented`. It looks like if the CodeUri were a relative path (the raw value in the template itself) instead of the absolute path when that method was called (i.e. `Resolving code path. Cwd=<MY_LOCAL_PATH>, CodeUri=./hello_world`), then it would properly resolve the absolute path to code and therefore work to use the `--docker-volume-basedir $PWD` option. This would allow for the functionality as it existed in prior versions of SAM CLI using that --docker-volume-basedir option and correspond to the documentation on the usage of that parameter which says: 

> If Docker is running on a remote machine, you must mount the path where the AWS SAM file exists on the Docker machine and modify this value to match the remote machine.

Requiring the configuration to use `working_dir` doesn't seem a viable resolution to #2492, especially in light of how it conflicts with the documentation around the `--docker-volume-basedir` parameter. It seems also related to: #963 and #1834. 

The desired behavior being described in this issue, #2492, #963 and #1834 seem like they would be resolved if you used  `--base-dir` with the value passed for `--docker-volume-basedir` or resolved the stack location from the outset using the `--docker-volume-basedir` parameter value.

**Comment by sbhvt** [comment]
@xazhao any updates on the above?

**Comment by andrewx82** [comment]
👍

**Comment by andrewx82** [comment]
@xazhao @hoffa  Could @sbhvt  and I potentially get an ETA on this resolution?

**Comment by andrewx82** [comment]
@xazhao @hoffa Could @sbhvt and I potentially get an ETA on this resolution?

**Comment by alexsanderp** [comment]
I solved this problem just using `network_mode: host` in my docker-compose.yml

**Comment by bfoura** [comment]
> I solved this problem just using `network_mode: host` in my docker-compose.yml

@alexsanderp Can you please share your docker-compose file ?

**Comment by alexsanderp** [comment]
file: docker-compose.yml
```
version: '3'
services:
    sam-api:
        image: default/sam-api:dev
        container_name: sam-api
        build:
            context: .
            dockerfile: sam.Dockerfile
        entrypoint: ["/app/sam-api-entrypoint.sh"]
        command: ["$PWD"]
        volumes:
            - .:/app
            - /var/run/docker.sock:/var/run/docker.sock
        ports:
            - "3000:3000"
        networks:
            - default

networks:
    default:
        name: default
 ```
 
 file: sam-api-entrypoint.sh
 ```
 #!/bin/sh

BASEDIR="$1"
sam build
sam local start-api \
  --host 0.0.0.0 \
  --docker-volume-basedir "${BASEDIR}" \
  --docker-network default \
  --debug
 ```
 
 So I can access the API from host with http:// localhost: 3000. But I can't access from other containers :(

**Comment by alexsanderp** [comment]
I was able to connect to the API from other containers using extra_hosts:

```
    extra_hosts:
      - host.docker.internal: host-gateway
```

Connect to API from inside the container:
`http://host.docker.internal:3000`

**Comment by andrewx82** [comment]
@alexsanderp We don't want to use volumes like your example is doing, we want a baked image that can be distributed

**Comment by andrewx82** [comment]
@xazhao any movement around this, it has been more than a month

**Comment by NickDarvey** [comment]
I get one step further by using the official SAM CLI images. Perhaps it's worth a try @andrewx82?

```yaml
version: '3.9'
services:
  api:
    image: public.ecr.aws/sam/build-nodejs14.x:1.26.0
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - ./dist:/var/task:ro
    ports:
      - 3000:3000
    command: sam local start-api --template stack.yaml --host 0.0.0.0 --docker-network application --container-host host.docker.internal --warm-containers EAGER

networks:
  default:
    name: application
```

Mind you, I'm now stuck because the spawned container's `/var/task` is empty even though it appears to correctly and successfully mount the volume from the CLI container. (#2955 may be related?) You might not run in to this depending on your environment. (I'm `Docker version 20.10.5, build 55c4c88` on Windows 10 in WSL 2 mode.)

```
api_1  | 2021-07-14 06:56:35  * Running on http://0.0.0.0:3000/ (Press CTRL+C to quit)
api_1  | Invoking app.handler (nodejs14.x)
api_1  | Skip pulling image and use local one: amazon/aws-sam-cli-emulation-image-nodejs14.x:rapid-1.26.0.
api_1  |
api_1  | Mounting /var/task/ResolveSource as /var/task:ro,delegated inside runtime container
api_1  | 2021-07-14T06:56:39.760Z       undefined       ERROR   Uncaught Exception      {"errorType":"Runtime.ImportModuleError","errorMessage":"Error: Cannot find module 'app'\nRequire stack:\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js","stack":["Runtime.ImportModuleError: Error: Cannot find module 'app'","Require stack:","- /var/runtime/UserFunction.js","- /var/runtime/index.js","    at _loadUserApp (/var/runtime/UserFunction.js:100:13)","    at Object.module.exports.load (/var/runtime/UserFunction.js:140:17)","    at Object.<anonymous> (/var/runtime/index.js:43:30)","    at Module._compile (internal/modules/cjs/loader.js:1085:14)","    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)","    at Module.load (internal/modules/cjs/loader.js:950:32)","    at Function.Module._load (internal/modules/cjs/loader.js:790:14)","    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)","    at internal/main/run_main_module.js:17:47"]}
```

**Comment by grahamlyons** [comment]
I tried @NickDarvey's `docker-compose.yaml` - I'm testing with a vanilla `sam init` hello world api, so I mounted `.` instead of `dist` - and I still get:

```
api_1  | No response from invoke container for HelloWorldFunction
api_1  | Invalid lambda response received: Lambda response must be valid json
api_1  | 2021-07-15 09:49:46 172.18.0.1 - - [15/Jul/2021 09:49:46] "GET /hello HTTP/1.1" 502 -
```

**Comment by Harriebo** [comment]
Hi, I ran into the same problem of running sam in a docker container for a project I work on. I manged in the end to fix it by running sam as:
sam local start-api -p 3001 --debug \
                                             --host 0.0.0.0 \
                                             --container-host 172.17.0.2 \
                                             --container-host-interface 0.0.0.0

notice that container-host is 172.17.0.2 because I override that in my docker daemon config, by default that would be 172.17.0.1 (on linux, on mac this is a hostname with a random ip I belive)
I also had to add -v "/var/run/docker.sock:/var/run/docker.sock:rw" do my docker run to make sure docker in docker works.

**Comment by xazhao** [comment]
@sbhvt @andrewx82 Sorry for the delay. 

@sbhvt Thanks for the detailed response and analysis. It did help me identify the bug a lot. When we introduced nested stack feature, we started resolving CodeUri before it is resolved using the value of `--docker-volume-basedir`. Therefore it won't be resolved with the directory you passed in even if you use that option. A fix in my mind is to remove the logic of first resolving.

@andrewx82 Could you please try the method that @sbhvt proposed using `--docker-volume-basedir` and `--base-dir`? In short, when `--base-dir` is used, it won't be resolved twice so that bug won't happen. For some reasons when I was running your example, it shows`Runtime.HandlerNotFound: dist/index.handler is undefined or not exported`. My guess is because of Mac zip issue but I haven't looked into that.

We will fix it once we confirm that is the root cause.

**Comment by grahamlyons** [comment]
I've tested this using the latest Sam CLI and an API Gateway project created from `sam init`:

`Dockerfile`:
```Dockerfile
FROM node:14-buster-slim

ENV NODE_ENV test

RUN apt-get update && \
    apt-get install -y \
        unzip \
        curl && \
    (cd /tmp/ && \
    curl -OL https://github.com/aws/aws-sam-cli/releases/download/v1.28.0/aws-sam-cli-linux-x86_64.zip && \
    unzip aws-sam-cli-linux-x86_64.zip -d sam-installation && \
    ./sam-installation/install)
```

`docker-compose.yaml`:
```yaml
version: '3.9'
services:
  sam_app:
    working_dir: $PWD
    build:
      context: .
      dockerfile: Dockerfile
    ports:
      - "3000:3000"
    command: sam local start-api --host 0.0.0.0 --container-host 172.17.0.1 --debug --docker-volume-basedir $PWD
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - $PWD:$PWD
```

This is the output I see from `docker-compose up`:
```shell
Creating network "sam-app_default" with the default driver
Creating sam-app_sam_app_1 ... done
Attaching to sam-app_sam_app_1
sam_app_1  | 
sam_app_1  |    SAM CLI now collects telemetry to better understand customer needs.
sam_app_1  | 
sam_app_1  |    You can OPT OUT and disable telemetry collection by setting the
sam_app_1  |    environment variable SAM_CLI_TELEMETRY=0 in your shell.
sam_app_1  |    Thanks for your help!
sam_app_1  | 
sam_app_1  |    Learn More: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-telemetry.html
sam_app_1  | 
sam_app_1  | 2021-08-18 10:22:04,409 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
sam_app_1  | 2021-08-18 10:22:04,410 | Using config file: samconfig.toml, config environment: default
sam_app_1  | 2021-08-18 10:22:04,410 | Expand command line arguments to:
sam_app_1  | 2021-08-18 10:22:04,410 | --template_file=/home/graham/workspace/sam-app/template.yaml --host=0.0.0.0 --container_host=172.17.0.1 --docker_volume_basedir=/home/graham/workspace/sam-app --port=3000 --static_dir=public --layer_cache_basedir=/root/.aws-sam/layers-pkg --container_host_interface=127.0.0.1 
sam_app_1  | 2021-08-18 10:22:04,493 | local start-api command is called
sam_app_1  | 2021-08-18 10:22:04,502 | No Parameters detected in the template
sam_app_1  | 2021-08-18 10:22:04,534 | 2 stacks found in the template
sam_app_1  | 2021-08-18 10:22:04,534 | No Parameters detected in the template
sam_app_1  | 2021-08-18 10:22:04,564 | 2 resources found in the stack 
sam_app_1  | 2021-08-18 10:22:04,564 | No Parameters detected in the template
sam_app_1  | 2021-08-18 10:22:04,595 | Found Serverless function with name='HelloWorldFunction' and CodeUri='hello-world/'
sam_app_1  | 2021-08-18 10:22:04,595 | --base-dir is not presented, adjusting uri hello-world/ relative to /home/graham/workspace/sam-app/template.yaml
sam_app_1  | 2021-08-18 10:22:04,600 | No Parameters detected in the template
sam_app_1  | 2021-08-18 10:22:04,630 | No Parameters detected in the template
sam_app_1  | 2021-08-18 10:22:04,659 | Found '1' API Events in Serverless function with name 'HelloWorldFunction'
sam_app_1  | 2021-08-18 10:22:04,659 | Detected Inline Swagger definition
sam_app_1  | 2021-08-18 10:22:04,659 | Lambda function integration not found in Swagger document at path='/hello' method='get'
sam_app_1  | 2021-08-18 10:22:04,659 | Found '0' APIs in resource 'ServerlessRestApi'
sam_app_1  | 2021-08-18 10:22:04,660 | Removed duplicates from '0' Explicit APIs and '1' Implicit APIs to produce '1' APIs
sam_app_1  | 2021-08-18 10:22:04,660 | 1 APIs found in the template
sam_app_1  | 2021-08-18 10:22:04,667 | Mounting HelloWorldFunction at http://0.0.0.0:3000/hello [GET]
sam_app_1  | 2021-08-18 10:22:04,667 | You can now browse to the above endpoints to invoke your functions. You do not need to restart/reload SAM CLI while working on your functions, changes will be reflected instantly/automatically. You only need to restart SAM CLI if you update your AWS SAM template
sam_app_1  | 2021-08-18 10:22:04,667 | Localhost server is starting up. Multi-threading = True
sam_app_1  | 2021-08-18 10:22:04  * Running on http://0.0.0.0:3000/ (Press CTRL+C to quit)
```

And when I hit `localhost:3000/hello` I see this output:
```shell
sam_app_1  | 2021-08-18 10:23:24,190 | Constructed String representation of Event to invoke Lambda. Event: {"body": null, "headers": {"Accept": "*/*", "Host": "localhost:3000", "User-Agent": "curl/7.64.0", "X-Forwarded-Port": "3000", "X-Forwarded-Proto": "http"}, "httpMethod": "GET", "isBase64Encoded": false, "multiValueHeaders": {"Accept": ["*/*"], "Host": ["localhost:3000"], "User-Agent": ["curl/7.64.0"], "X-Forwarded-Port": ["3000"], "X-Forwarded-Proto": ["http"]}, "multiValueQueryStringParameters": null, "path": "/hello", "pathParameters": null, "queryStringParameters": null, "requestContext": {"accountId": "123456789012", "apiId": "1234567890", "domainName": "localhost:3000", "extendedRequestId": null, "httpMethod": "GET", "identity": {"accountId": null, "apiKey": null, "caller": null, "cognitoAuthenticationProvider": null, "cognitoAuthenticationType": null, "cognitoIdentityPoolId": null, "sourceIp": "192.168.112.1", "user": null, "userAgent": "Custom User Agent String", "userArn": null}, "path": "/hello", "protocol": "HTTP/1.1", "requestId": "313b059f-6e74-4d2d-9033-d47f935ca834", "requestTime": "18/Aug/2021:10:22:04 +0000", "requestTimeEpoch": 1629282124, "resourceId": "123456", "resourcePath": "/hello", "stage": "Prod"}, "resource": "/hello", "stageVariables": null, "version": "1.0"}
sam_app_1  | 2021-08-18 10:23:24,191 | Found one Lambda function with name 'HelloWorldFunction'
sam_app_1  | 2021-08-18 10:23:24,191 | Invoking app.lambdaHandler (nodejs14.x)
sam_app_1  | 2021-08-18 10:23:24,191 | No environment variables found for function 'HelloWorldFunction'
sam_app_1  | 2021-08-18 10:23:24,191 | Environment variables overrides data is standard format
sam_app_1  | 2021-08-18 10:23:24,191 | Loading AWS credentials from session with profile 'None'
sam_app_1  | 2021-08-18 10:23:26,219 | Resolving code path. Cwd=/home/graham/workspace/sam-app, CodeUri=/home/graham/workspace/sam-app/hello-world
sam_app_1  | 2021-08-18 10:23:26,219 | Resolved absolute path to code is /home/graham/workspace/sam-app/hello-world
sam_app_1  | 2021-08-18 10:23:26,220 | Code /home/graham/workspace/sam-app/hello-world is not a zip/jar file
sam_app_1  | 2021-08-18 10:23:26,275 | Skip pulling image and use local one: public.ecr.aws/sam/emulation-nodejs14.x:rapid-1.28.0.
sam_app_1  | 
sam_app_1  | 2021-08-18 10:23:26,275 | Mounting /home/graham/workspace/sam-app/hello-world as /var/task:ro,delegated inside runtime container
sam_app_1  | 2021-08-18 10:23:27,132 | Starting a timer for 1 seconds for function 'HelloWorldFunction'
sam_app_1  | 2021-08-18 10:23:28,866 | Cleaning all decompressed code dirs
sam_app_1  | 2021-08-18 10:23:28,866 | No response from invoke container for HelloWorldFunction
sam_app_1  | 2021-08-18 10:23:28,866 | Invalid lambda response received: Lambda response must be valid json
sam_app_1  | 2021-08-18 10:23:28 192.168.112.1 - - [18/Aug/2021 10:23:28] "GET /hello HTTP/1.1" 502 -
```

From @sbhvt's version I had to add in the `working_dir` line otherwise I got the error `Error: Template file not found at /var/opt/template.yml`.

When I add the `--base-dir` option I get the error: `Error: no such option: --base-dir`.

**Comment by metaskills** [comment]
HUGE THANKS to @sbhvt's comment here https://github.com/aws/aws-sam-cli/issues/2837#issuecomment-845487064

I've been working on another bug/feature needed for easy code reloading when using containers (https://github.com/aws/aws-sam-cli/issues/921#issuecomment-907859353) and the `working_dir` information helped me in a hack workaround for my issue.... but please PLEASE SAM team address this issue.

**Comment by grahamlyons** [comment]
Is there any update on this?

**Comment by ann0nc0d3r** [comment]
This has completely blocked any sort of development whatsoever. I've been trying to get this working for the best part of 3 days and have got nowhere. In order to test any changes I'm having to deploy the code and hope for the best 🤷🏻‍♂️

This should be a much higher priority than it seems to be at the moment.

**Comment by grahamlyons** [comment]
@ann0nc0d3r, it looks like this is being worked on at the moment in #3237. There are commits within the last 24 hours so hopefully it'll be fixed imminently 🤞🏻

**Comment by rdallasgray** [comment]
Appears #3237 was merged, released and then reverted without explanation -- is a fix still ongoing?

**Comment by jfuss** [comment]
I am trying to understand this issue but failing to grasp what is going on.

I will try to do some testing but want to ask some clarifying questions to the group here.

1. What use-case does docker-compose solve?
2. Why not install SAM CLI on your system and use as documented?
3. Can someone provide a full working example (with steps) of this issue (there are parts throughout but its unclear if there are multiple issues going on or not)?

**Comment by metaskills** [comment]
> What use-case does docker-compose solve?

We use the SAM build images as development images. Docker compose is a nice way to spin up that service and the other(s) the function may need. Example: https://lamby.custominktech.com/docs/anatomy#our-sam-cookiecutters-features

> Why not install SAM CLI on your system and use as documented?

Cause dev containers are the way! Simplifies tooling, etc.

> Can someone provide a full working example

You can use the quick start here? https://lamby.custominktech.com/docs/quick_start

**Comment by grahamlyons** [comment]
> Appears #3237 was merged, released and then reverted without explanation -- is a fix still ongoing?

Yes, it would be really nice to understand why the proposed fix was reverted.

**Comment by jfuss** [comment]
@metaskills What I have a hard time with is, running SAM CLI in a container seems like the wrong experience. I know it is there to enable building in CI but testing locally is a different story. The reason being is SAM CLI needs to talk on localhost so if you want to deal with all kinds of paths and networking you can, but it is MUCH easier to install and run SAM CLI locally. That is how it was originally intended to run. 

If people want to run through docker-compose, you can but their be dragons. Still trying to understand what is going on here and from the original post is smells of docker networking.

@grahamlyons I am talking with the team but from what I can gather from this issue is that their are likely two things going on. The "proposed fix" seems to help with nested stacks. The original post and example provided are not using nested stacks, so (at this time) asking about the revert is distracting. Let's first get an understanding of the original issue and validate. We can come back to the revert later. It's possible what fix will help here too but I want to keep things separate and focused.

What likely happened is the changes caused impact to our tests we run on releases and needed to be reverted so that it was not blocking whatever was going out. We typically don't revert changes unless we have to and elect to roll forward as much as possible when issues are found.

**Comment by metaskills** [comment]
> I know it is there to enable building in CI but testing locally is a different story.

But that is not the full story. The idea of the development container is that it also the CI/CD container. So when you run your tests locally, it is the same platform, docker setup, etc being used by CircleCI/GitHubActions for CI tests. Same is true for CD... the same docker compose setup is used to on CircleCI/GitHubActions to build and deploy via SAM. No one every runs a deploy from local dev. Does that help explain how development containers crosses boundaries and uses cases better? IE. CI/CD uses the same containers and that is super powerful?

**Comment by metaskills** [comment]
In fact, the emphasis is on CI/CD first... development is the benefit of a good CI/CD container setup. Again, that project links demos this all well. Also, I'm writing up a big post on using CI/CD containers for development with systems like GitHub Codespaces or GitPod. I can share that later but hope this info now helps?

**Comment by jfuss** [comment]
@metaskills I think it a "to each their own" thing. I personally don't think that is ideal. I much rather run tooling locally. Because of how SAM CLI works and interacts with docker, you get into a bunch of weird things when trying to use parts of the CLI that also interact with Docker. Not to say my opinion is right, but it is my stance. Please do share, I would like to understand the draw to these setups more. Not only for my own learning/curve opinions but to also get a better understand of what customers do to help inform the tooling we build.

Sorry for the rant/opinions, trying not to get my opinions and preferences in the way of the issue.

I am still trying to validate the original post. It feels and looks like docker networking but I recognize there could be another issue here too. So trying to keep things separate to help me focus on root causing the right things (which might be two things).

**Comment by metaskills** [comment]
Thanks! I will share that my feedback is not personal but based on using SAM well in a semi large organization using both CircleCI & GitHub Actions. Years ago we had massive tooling pains between Engineering, DX, and Operation teams. Installing SAM locally has gotten better with binaries but it has been nice to standardize on Linux and containers and solved a lot of teams to use our CI/CD pipelines. Gone are the days of building scripts for different platforms for local dev and different host interfaces to CI/CD. 

Example, when engineers need to update SAM CLI, they simply change this docker tag. This is one of many examples tho. https://github.com/customink/lamby-cookiecutter/blob/master/%7B%7Bcookiecutter.project_name%7D%7D/Dockerfile-build#L1

**Comment by jfuss** [comment]
@metaskills I can understand the draw but the pain isn't worth it to me. Docker in Docker is a terrible experience, especially in how SAM CLI does things (communicating with localhost).

So in talking with the team, the original post was addressed but another issue came up (around the nested stack/how directory resolving works). We did revert the previous change because of unexpected failures/impact and have not picked back up the issue. 

We don't provide ETA on issues but we are prioritizing this within the engineering team. We will reach back out to this group if we do need some help or validation on the solution. We do tag issues in our PRs, so there will be a link to that once the team takes this up. Thanks for being patient with us on getting this addressed.

**Comment by awilkins** [comment]
Having issue with the same symptoms

First, I had to resolve the issue where SAM is binding Lambda executors to both the `bridge` network and your stated `docker-network`, which stops them resolving other containers by name (e.g. `localstack` for Dynamo and SSM). The Docker CLI refuses to let you do this, which I think is an indication that it's not a good configuration.

However, by using fixed IP addresses and passing them to Lambdas in env variables, I have overcome this.

But... same error

```
agent_1  | api_1         | 2021-12-01 11:56:17,674 | No response from invoke container for health
agent_1  | api_1         | 2021-12-01 11:56:17,674 | Invalid lambda response received: Lambda response must be valid json
```

If you open a shell in the container running `sam local start-api`  ... and work out which container is running your Lambda

```
root@e62d41a5eb57:/src/validation# curl -XPOST http://172.24.0.5:8080/2015-03-31/functions/function/invocations -d'{}'
{"statusCode": 200, "body": "OK"}
```

So on the face of it, it ought to work.

Hmm. <splices a local copy of `samcli` package with a little extra logging into container>

```
2021-12-01 15:53:05,102 | Starting a timer for 3 seconds for function 'health'
agent_1  | api_1         | 2021-12-01 15:53:05,102 | Invoke : http://api:5936/2015-03-31/functions/function/invocations
                                                                    |   |    |
                                                                     --- ----
                                                                       |   |
                                                                       |ARRRRGH
                                                                       |- also arrrgh
```

So : 

- It's trying to invoke the function by calling into the same container running the API
- And using the remapped port

This works when `sam` is running on host, because those ports are exposed at `localhost`.  
This does not work when `sam` is in a stack (on a user network) because the functions are not on the same container and the remapped port is not available outside.

In `local_lambda.py` this patch seems to resolve it

```python
# @ 187
        if self.network_id and self.network_id != "host":
            try:
                # Disconnect the container from the bridge so name resolution works
                bridge_network = self.docker_client.networks.get("bridge")
                bridge_network.disconnect(self.id)
                network = self.docker_client.networks.get(self.network_id)
                network.connect(self.id)
                self._container_host = real_container.id[:12]
                self.rapid_port_host = self.RAPID_PORT_CONTAINER
            except DockerNetworkNotFound:
                # stop and delete the created container before raising the exception
                real_container.remove(force=True)
                raise
```

**Comment by awilkins** [comment]
Have started PR #3509 but feel it might need a little work.

**Comment by awilkins** [comment]
Indeed, you're right, was quoting the wrong tab header in my editor : see PR with patch in this stretch of code

**Comment by grahamlyons** [comment]
@jfuss, I've been a big fan of virtualised development environments for many years now. I used VirtualBox for a long time and ran a Linux distro of whatever was being used in production. I found that it meant there were a lot of problems I only had to solve once, like installing binary dependencies. (It's also meant that I've not bricked my host machine by mistake a couple of times).

More recently, containers provide an excellent way to codify test requirements and dependencies. My aim for any project I work on is to have a single command to run the tests from a fresh clone of the codebase. This lowers the bar to contributing to about as low as is feasible.

With the projects that I'm using AWS Sam for they usually have a dependency like a DynamoDB table or an S3 bucket so being able to put those things into a `docker-compose.yml` file and have them automatically pulled and initialised is amazing - the dependencies and their config are documented in the file that drives them.

With the [Remote Containers plugin](https://code.visualstudio.com/docs/remote/containers-tutorial) for VSCode it even lets you drop into a terminal in one of those containers so that you can develop inside it. If you're running code inside a container in production then you can use the exact same image for developing in - a slight aside to this thread because that's not the way that Sam works.

The point of the evangelism above is to let you know that lots of people (whom I've indoctrinated) use containers to run everything locally because, despite some headaches that you have very accurately pointed out, they solve a lot of other headaches.

I have managed to get a Lambda/API Gateway/DynamoDB project using Sam running using Docker Compose, with acceptance tests running from another container and everything talking to each other. While I was typing this message I've just updated to the latest Sam (1.36.0) and it is still working so I will strip down to a minimal example and put it in a repo for anyone on this thread who is interested.

**Comment by grahamlyons** [comment]
https://github.com/grahamlyons/aws-sam-docker-compose-example

Here is an example of using the latest AWS Sam in a project using Lambda, API Gateway and DynamoDB, with Docker Compose handling local dependencies. I've tested this on macOS and Linux but Windows.

I would love feedback on the project set up and whether it works for people who have had problems with this. There's quite a bit of complexity in there, which I'm not keen on, but I like the ease of running tests with local dependencies from a fresh checkout - being able to drop that same script into CI - I think - is a big win.

**Comment by jfuss** [comment]
@metaskills @grahamlyons I wonder if we should move this discussion more to a Github Discussion post? I don't think it is enabled in the repo right now (which I can work on) but could give us a platform to discuss topics like this without "rabbit hole-ing" on these issues. Probably worth it in some cases but don't want to distract from the issue. It makes following/understanding status harder. 

I think there is value for some things, I agree there. I find the struggle with docker in docker not worth it at all (just tradeoffs at the end of the day). I prefer testing on AWS rather than local for as much as possible and generally think local is great for isolating testing but running locally gives false confidence (docker container != Lambda sandbox, might be close enough for some things but seen too many "why is local different").

**Comment by grahamlyons** [comment]
@jfuss, that sounds like a great idea. I'd love to say more on this topic and if you'd prefer to move it out of this thread then I'll happily stand on another soap box!

**Comment by grahamlyons** [comment]
Just to summarise on what is relevant for this issue: I think my example demonstrates that it is possible to use AWS Sam in Docker Compose environment and get a response from the containers. That being said, it's definitely not simple and it took me a long time without many examples to work from.

I'd like the use case to be considered by the Sam team and ideally made easier but at the very least not taken away.

**Comment by wberrytbt** [comment]
@grahamlyons ur example helped unblock me.  I confirmed that your test example (well, not all the tests) work on Linux.  The root endpoint test I did get working.  You might be able to construct a more minimal example without the second acceptance test, just the root handler (or a hello world example).

But thanks anyway!  I think docker Linux added recently the same extra hosts mapping, so Mac and Linux should both work in my experience.

**Comment by grahamlyons** [comment]
@wberrytbt, the example could have been simpler if it hadn't included DynamoDB but I wanted to demonstrate the Lambda container talking to a service it depends on.

If it doesn't all work from a fresh clone when you run `./test.sh` then please let me know what went wrong. You can see it working on the Ubuntu environment that GitHub Actions provides, and I've run it successfully locally on Mac and Linux.

**Comment by wberrytbt** [comment]
@grahamlyons one issue was the docker network (was not defined).  You can specify it as external in the docker-compose yaml and add a create network command in test.sh if you want to fix that specific issue.  

Honestly, after I got test #1 passing, did not invest the time to get #2 working (since serving hello-world docker in docker was the issue I intended to solve).  I can spend some time Monday if that adds value, and do see your point in connecting another service (and also like your test example), but if you wanted a truly minimal example-- the second test and dynamo is unnecessary.  But I appreciate it, either way!

**Comment by grahamlyons** [comment]
> one issue was the docker network (was not defined). You can specify it as external in the docker-compose yaml and add a create network command in test.sh if you want to fix that specific issue.

That's not necessary in the general case; you can see it running in GitHub Actions without having to do that:
https://github.com/grahamlyons/aws-sam-docker-compose-example/runs/4408692591?check_suite_focus=true

If you figure out why you had to do that then let me know.

**Comment by rocky112358** [comment]
Hi, it would be nice if you could let me know where the recent conversation about running SAM CLI in a container (virtualize d environment) is being continued. I am interested to the topic, too.

**Comment by rocky112358** [comment]
@jfuss @grahamlyons @metaskills Hi, it would be nice if you could let me know where the recent conversation about running SAM CLI in a container (virtualize d environment) is being continued. I am interested to the topic, too.

**Comment by grahamlyons** [comment]
> @jfuss @grahamlyons @metaskills Hi, it would be nice if you could let me know where the recent conversation about running SAM CLI in a container (virtualize d environment) is being continued. I am interested in the topic, too.

There's nothing organised yet as far as I'm aware. I too am keen to hear more from the Sam team and from users with similar use cases.

**Comment by grahamlyons** [comment]
@jfuss, is there anywhere that we can continue this discussion?

**Comment by austinben** [comment]
Also interested about this. 

I have success running the commands locally, but when I run them in CircleCI, I am faced with the dreaded `No response from invoke container`. Is there special flags to provide to the `sam` commands for running in the circleCI environment to allow communication between the container and the rest of the environment?  I assume that is the issue.

I am using the following two orbs:
```
  aws-cli: circleci/aws-cli@2.0.6
  sam: circleci/aws-sam-serverless@3.2.0
 ```

**Comment by grahamlyons** [comment]
@austinben, what OS are you running on locally? In the example I posted above I got it working on macOS but then had to add something else to make it work on Linux. I don't remember exactly what but take a look and see if you've anything missing.

**Comment by jfuss** [comment]
@grahamlyons Sorry for my slow response. Was on leave for a while at the end of last year/beginning of this year and then this just fell off my radar. 

There was some back and forth whether Discussions was useful. For now, a lightweight way (I can think of) is just move discussion about "docker in docker" into a [Discussion] titled issue. That will at least allow that to focus. We do have a slack channel, which could be another option.

**Comment by grahamlyons** [comment]
@jfuss, no worries - thanks for coming back on this.

I would welcome any discussion on running Sam with Docker (although I'd avoid the phrase "docker in docker" as that implies running the docker daemon inside a container which is whole other thing) as it's something I've been doing for a while so would be good to air my use cases.

**Comment by leoGalani** [comment]
btw,
Once you define the docker network, you don't need to use the `host.docker.internal` anymore.
You can use the service name on the URL you want to connect :)

**Comment by hawflau** [comment]
Closing as the fix has been released in SAM v1.57.0

**Comment by awilkins** [comment]
Hey, @hawflau ; I'm not convinced this issue is the one fixed by the PR merged above.

That PR seemd to resolve Issue 3520 (not going to hotlink it to avoid associating it with this ticket).

This ticket is to do with accessing APIs over a Docker user network. I'm not sure why that PR mentions this issue by ID.

**Comment by lucashuy** [comment]
Hi, thanks for bringing up your observations! I can confirm that the linked PR fixed one of the original issues (not being able to target the correct directory on the host machine). It looks like there is another issue related to the usage of the `--docker-network` flag in `sam local`. After testing with the OP's original files and removing the `--docker-network host` flag, I'm able to invoke the API using `curl` on the host machine.

We'll investigate further and look into next steps to resolve this.

**Comment by tariromukute** [comment]
Hi, the thread was helpful, I managed to get the app working, however, I started facing issues when I was trying to use layers. My docker-compose file is as below.

```
version: '3'
services:
  lambdas:
    build:
      context: ./
      dockerfile: Dockerfile
    image: myapp-lambda:local
    ports:
      - 3000:3000
    environment:
      - SAM_CLI_CONTAINER_CONNECTION_TIMEOUT=12
      # This will override the AWSEndpoint paramter defined in the template
      - AWS_ENDPOINT=http://host.docker.internal:4566 # localstack container host for dynamodb
      - AWS_DEFAULT_REGION='us-west-2'
      - AWS_REGION='us-west-2'
      - AWS_ACCESS_KEY_ID='fake-access-key-id'
      - AWS_SECRET_ACCESS_KEY='fake-secret-key'
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./:/var/task
    entrypoint: sam local start-api --host 0.0.0.0 --port 3000 --docker-volume-basedir $PWD --container-host host.docker.internal

  localstack:
    image: localstack/localstack:latest
    ports:
      - '8000:8000' # using port 8000 to be consistent with dynamodb local jar
      - '4566:4566' # new dynamodb port
      - '8080:8080' # the localstack admin portal
    container_name: dynamodb_localstack
    environment: 
      - AWS_DEFAULT_REGION=us-east-1
      - EDGE_PORT=4566
      - SERVICES=dynamodb,s3
      # App uses different profiles, this allows a single database file will be shared for all clients
      - DYNAMODB_SHARE_DB=1
```
The error I am getting is that the path to the utilities-layer is not being found. The path is on the host machine and it exists. I am running a Mac OS M1 using a docker desktop. See the error I am facing below.

```
FileNotFoundError: [Errno 2] No such file or directory: '/Users/tariromukute/Documents/dev/app/utilities-layer'
```
I tried to add the flag `--layer-cache-basedir=/root/.aws-sam/layers-pkg` with no luck. Anyone with an idea to get the sam app working with docker-compose when using layers? See a section of the template.yaml below.

```
Resources:  
....

  utilitiesLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
      LayerName: utilities-layer
      Description: Utility layer
      ContentUri: utilities-layer/
      CompatibleRuntimes:
        - nodejs12.x
      LicenseInfo: "MIT"
      RetentionPolicy: Retain

Globals:
  Function:
    Runtime: nodejs14.x
    Architectures:
      - arm64
    MemorySize: 128
    Timeout: 100
    Layers:
      - !Ref utilitiesLayer
```

**Comment by zahid-hussain-dev** [comment]
Any resolution? I am also encountering the same issue with SAM.

**Comment by awilkins** [comment]
The [PR linked above](https://github.com/aws/aws-sam-cli/pull/3509) seems to fix it, but hasn't been merged, partly because I stopped working with this particular set of components and didn't have time to work on it. I have no idea if the code has diverged enough to break the patch.
